// client/app/stores/PlannerStore.js
import pickBy from 'lodash/pickBy';
import moment from 'moment';
import alt from 'alt-instance';

import PlannerActions from 'actions/PlannerActions';
import PlannerSource from 'sources/PlannerSource';
import GroceryListSource from 'sources/GroceryListSource';


class PlannerStore {
  constructor() {
    this.date = moment().startOf('day');
    this.planner = [];
    this.modal = {
      recipe: null,
      date: null,
      meal: null,
      open: false,
    };

    this.bindListeners({
      handleFetchPlanner: PlannerActions.FETCH_PLANNER,
      handleChangeDate: PlannerActions.CHANGE_DATE,
      handleAddRecipe: PlannerActions.ADD_RECIPE,
      handleGenerateGroceries: PlannerActions.GENERATE_GROCERIES,
      handleShowModal: PlannerActions.SHOW_MODAL,
      handleCloseModal: PlannerActions.CLOSE_MODAL,
    });
  }

  handleFetchPlanner() {
    PlannerSource.fetch(this.date.format('YYYY-MM-DD'))
      .then(({data}) => {
        this.setState({planner: data});
      });
  }

  handleChangeDate(date) {
    this.setState({date});
    PlannerActions.fetchPlanner.defer();
  }

  handleAddRecipe({recipe, date, meal}) {
    PlannerSource.addEntry(date, {meal, recipe: recipe._id})
      .then(({data}) => {
        const date = moment(data.date).startOf('day');
        const index = this.planner.findIndex(plan => moment(plan.date).startOf('day').isSame(date));
        this.setState({
          planner: [...this.planner.slice(0, index), data, ...this.planner.slice(index + 1)]
        });
      });
  }

  handleShowModal(props) {
    this.setState({modal: {...this.modal, ...pickBy(props, x => x), open: true}});
  }

  handleCloseModal() {
    this.setState({modal: {open: false}});
  }

  handleGenerateGroceries(entry) {
    const ingredients = entry.recipe.ingredients.map((i) => {
      return {
        name: i.name,
        completed: false
      };
    });
    GroceryListSource.create({
      title: `${entry.meal}: ${entry.recipe.title}`,
      items: ingredients
    });
  }
}

export default alt.createStore(PlannerStore, 'PlannerStore');
