// client/app/stores/RecipeStore.js
import {v4 as uuid} from 'node-uuid';
import flow from 'lodash/fp/flow';
import clone from 'lodash/clone';
import values from 'lodash/fp/values';
import set from 'lodash/fp/set';
import unset from 'lodash/fp/unset';
import omit from 'lodash/fp/omit';
import RecipeActions from 'actions/RecipeActions';
import RecipeSource from 'sources/RecipeSource';
import {nonEmptyIngredient} from 'helpers/IngredientHelper';
import {history} from 'routes';
import alt from '../alt';

const createIngredient = () => ({_id: uuid()});

const stripIds = (ingredients) => ingredients.map(ingredient => omit('_id')(ingredient));
const stripEmptyIngredients = (ingredients) => ingredients.filter(nonEmptyIngredient);

class RecipeStore {
  static config = {
    getState(currentState) {
      return {
        recipe: {
          ...currentState.recipe,
          ingredients: values(currentState.ingredients)
        }
      };
    }
  };

  constructor() {
    this.recipe = {
      title: ''
    };
    this.ingredients = {};
    this.err = null;

    this.bindListeners({
      handleFetchRecipe: RecipeActions.FETCH_RECIPE,
      handleFetchSuccess: RecipeActions.FETCH_SUCCESS,
      handleFetchFailure: RecipeActions.FETCH_FAILURE,
      handleLoadRecipe: RecipeActions.LOAD_RECIPE,
      handleUpdateRecipe: RecipeActions.UPDATE_RECIPE,
      handleSubmitRecipe: RecipeActions.SUBMIT_RECIPE,
      handleSubmitSuccess: RecipeActions.SUBMIT_SUCCESS,
      handleSubmitFailure: RecipeActions.SUBMIT_FAILURE,
      handleAddIngredient: RecipeActions.ADD_INGREDIENT,
      handleRemoveIngredient: RecipeActions.REMOVE_INGREDIENT,
      handleUpdateIngredient: RecipeActions.UPDATE_INGREDIENT,
    });
  }

  handleFetchRecipe(recipeId) {
    RecipeSource.get(recipeId);
  }

  handleFetchSuccess({result, entities}) {
    this.setState({recipe: entities.recipes[result], err: null});
  }

  handleFetchFailure(err) {
    this.setState({err})
  }

  handleLoadRecipe(recipe) {
    this.setState({recipe});
  }

  handleUpdateRecipe({key, value}) {
    this.setState({recipe: {...this.recipe, [key]: value}});
  }

  handleSubmitRecipe() {
    RecipeSource.create({
      ...this.recipe,
      ingredients: flow(
        values,
        stripIds,
        stripEmptyIngredients,
      )(this.ingredients)
    });
  }

  handleSubmitSuccess() {
    history.push('/recipes');
  }

  handleSubmitFailure(err) {
    this.setState({err});
  }

  handleAddIngredient() {
    const withId = createIngredient();
    this.setState({ingredients: set(withId._id, withId)(this.ingredients)});
  }

  handleRemoveIngredient(ingredientId) {
    this.setState({ingredients: unset(ingredientId)(this.ingredients)});
  }

  handleUpdateIngredient({ingredientId, key, value}) {
    const updatedIngredient = flow(
      clone,
      set(key, value)
    )(this.ingredients[ingredientId]);
    this.setState({ingredients: set(ingredientId, updatedIngredient)(this.ingredients)});
  }
}

export default alt.createStore(RecipeStore, 'RecipeStore');
