// client/app/stores/RecipeStore.js
import assign from 'lodash/assign';
import values from 'lodash/values';
import RecipeActions from 'actions/RecipeActions';
import RecipeSource from 'sources/RecipeSource';
import alt from '../alt';

class RecipeStore {
  static config = {
    getState(currentState) {
      return assign({}, currentState, {recipes: values(currentState.recipes)});
    }
  };

  constructor() {
    this.recipes = {};
    this.err = null;
    this.loading = false;

    this.bindListeners({
      handleFetchLists: RecipeActions.FETCH_RECIPES,
      handleFetchSuccess: RecipeActions.FETCH_SUCCESS,
      handleFetchFailure: RecipeActions.FETCH_FAILURE,
    });
  }

  handleFetchLists() {
    if (!this.loading) {
      this.setState({loading: true});
      RecipeSource.fetch();
    }
  }

  handleFetchSuccess({entities}) {
    this.setState({recipes: entities.recipes, err: null, loading: false});
  }

  handleFetchFailure(err) {
    this.setState({err, loading: false});
  }
}

export default alt.createStore(RecipeStore, 'RecipeStore');
