// client/app/components/foodlog/AddMealContainer.jsx
import React from 'react';

import AddMeal from './AddMeal';

const INITIAL_STATE = {selectedIngredient: {name: '', ndbno: '-1'}};

export default class AddMealContainer extends React.Component {
  constructor(...args) {
    super(...args);

    this.state = INITIAL_STATE;

    this.valueChanged = this.valueChanged.bind(this);
    this.buttonClicked = this.buttonClicked.bind(this);
  }

  valueChanged(value) {
    if(value) {
      this.setState({selectedIngredient: {name: value.name, ndbno: value.ndbno}});
    } else {
      this.setState(INITIAL_STATE);
    }
  }

  buttonClicked() {
    if(this.state.selectedIngredient.name) {
      this.props.onAdd(this.state.selectedIngredient);
    }
    this.setState(INITIAL_STATE);
  }

  render() {
    return (
      <AddMeal value={this.state.selectedIngredient.name}
                placeholder={this.props.placeholder}
                onChange={this.valueChanged}
                onSubmit={this.buttonClicked} />
    );
  }
}

AddMealContainer.propTypes = {
  placeholder: React.PropTypes.string,
  onAdd: React.PropTypes.func,
};

AddMealContainer.defaultProps = {
  placeholder: '',
  onAdd: (value) => {}
};
