// client/app/components/foodlog/AddMeal.jsx
import React from 'react';
import {FormGroup, InputGroup, FormControl, Button, Glyphicon} from 'react-bootstrap';
import IngredientsAutocomplete from 'client/app/components/shared/IngredientsAutocomplete';

import './AddMeal.scss';

const INITIAL_STATE = {selectedIngredient: {name: '', ndbno: '-1'}};

export default class AddMealContainer extends React.Component {
  constructor(...args) {
    super(...args);

    this.state = INITIAL_STATE;

    this.valueChanged = this.valueChanged.bind(this);
    this.buttonClicked = this.buttonClicked.bind(this);
  }

  valueChanged(value) {
    if(value) {
      this.setState({selectedIngredient: {name: value.name, ndbno: value.ndbno}});
    } else {
      this.setState(INITIAL_STATE);
    }
  }

  buttonClicked() {
    if(this.state.selectedIngredient.name) {
      this.props.onAdd(this.state.selectedIngredient);
    }
    this.setState(INITIAL_STATE);
    this.refs.autocomplete.clearSelection();
  }

  render() {
    return (
      <FormGroup className="s__AddMeal">
        <InputGroup>
          <IngredientsAutocomplete ref="autocomplete" onIngredientSelected={this.valueChanged}/>
          <InputGroup.Button>
            <Button onClick={this.buttonClicked} aria-label="Add"><Glyphicon glyph="plus" /></Button>
          </InputGroup.Button>
        </InputGroup>
      </FormGroup>
    );
  }
}

AddMealContainer.propTypes = {
  placeholder: React.PropTypes.string,
  onAdd: React.PropTypes.func,
};

AddMealContainer.defaultProps = {
  placeholder: '',
  onAdd: (value) => {}
};
