// client/app/components/foodlog/FoodLog.jsx
import React from 'react'

import storeStateSubscriberCreator from 'client/app/mixins/StoreStateSubscriber';

import MealsStore from 'client/app/stores/MealsStore';
import MealsCrudActions from 'client/app/actions/MealsCrudActions';

import ListGroup from 'react-bootstrap/lib/ListGroup';
import ListGroupItem from 'react-bootstrap/lib/ListGroupItem';
import AddInput from 'client/app/components/shared/form/AddInput';

export default class FoodLogView extends React.Component {
  constructor(...props) {
    super(...props);
    this.state = {
      newMealName: ''
    };
  }

  render() {
    return (
      <div>
        <h2>Food Log</h2>
        <ListGroup>
          {this.props.meals.map(m => (<ListGroupItem key={m._id}>{m.name}</ListGroupItem>))}
        </ListGroup>
        <AddInput 
          placeholder="Add meal"
          onChange={(newValue => this.setState({newMealName: newValue})).bind(this)}
          onSubmit={((name) => {
            this.props.onMealAdded(this.state.newMealName);
            this.setState({newMealName: ''});
          }).bind(this)}
        />
      </div>
    )
  }
}

const MealsStoreStateSubscriber = storeStateSubscriberCreator(MealsStore);
export default class FoodLogViewContainer extends MealsStoreStateSubscriber(React.Component) {
  componentDidMount() {
    super.componentDidMount();
    MealsCrudActions.fetchMeals();
  }

  render() {
    return (
      <FoodLogView
        meals={this.state.meals}
        onMealAdded={(name) => MealsCrudActions.addMeal(name)}
      />
    );
  }
}
