// client/app/components/groceries/list/AddItemForm.jsx
import {v4 as uuid} from 'node-uuid';
import assign from 'lodash/assign';
import React from 'react';
import GroceryListSchema from 'schemas/GroceryListSchema';
import GroceryListActions from 'actions/GroceryListActions';
import AddInput from 'components/shared/form/AddInput';

const createItem = () => ({_id: uuid(), name: ''});

export default class AddItemForm extends React.Component {
  constructor(...args) {
    super(...args);

    this.state = {item: createItem()};

    this.handleNameChange = this.handleNameChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleNameChange(name) {
    this.setState({item: assign({}, this.state.item, {name})});
  }

  handleSubmit() {
    GroceryListActions.addItem(this.props.list._id, this.state.item);
    this.setState({item: createItem()});
  }

  render() {
    return (
      <AddInput value={this.state.item.name}
                placeholder="Add item..."
                onChange={this.handleNameChange}
                onSubmit={this.handleSubmit} />
    );
  }
}

AddItemForm.propTypes = {
  list: GroceryListSchema.only('_id').isRequired,
};
