// client/app/components/groceries/list/GroceryListContainer.jsx
import assign from 'lodash/assign';
import React from 'react';
import GroceryListActions from 'actions/GroceryListActions';
import ListSchema from 'schemas/GroceryListSchema';

import GroceryList from './GroceryList';

class ListContainer extends React.Component {
  constructor(...args) {
    super(...args);

    this.state = {item: {name: '', completed: false}};

    this.handleItemChange = this.handleItemChange.bind(this);
    this.handleItemSubmit = this.handleItemSubmit.bind(this);
    this.handleItemToggle = this.handleItemToggle.bind(this);
    this.handleItemDelete = this.handleItemDelete.bind(this);
  }

  handleItemChange(name) {
    this.setState({item: assign({}, this.state.item, {name})});
  }

  handleItemSubmit() {
    GroceryListActions.addItem(this.props.list._id, this.state.item);
  }

  handleItemToggle(itemId) {
    GroceryListActions.toggleItem(this.props.list._id, itemId);
  }

  handleItemDelete(itemId) {
    GroceryListActions.removeItem(this.props.list._id, itemId);
  }

  render() {
    const {list} = this.props;
    return (
      <GroceryList list={list}
                   onItemChange={this.handleItemChange}
                   onItemSubmit={this.handleItemSubmit}
                   onItemToggle={this.handleItemToggle}
                   onItemDelete={this.handleItemDelete} />
    );
  }
}

ListContainer.propTypes = {
  list: ListSchema.only('_id').isRequired
};

export default ListContainer;
