// server/api/recipe/recipe.controller.js
/**
 * GET     /api/recipes              ->  index
 * POST    /api/recipes              ->  create
 * GET     /api/recipes/:id          ->  show
 * PUT     /api/recipes/:id          ->  update
 * DELETE  /api/recipes/:id          ->  destroy
 */

import {respondWithResult, saveUpdates, handleEntityNotFound, handleError} from 'server/api/common';
import Recipe from './recipe.model';

function removeEntity(res) {
  return function(entity) {
    if(entity) {
      return entity.removeAsync()
        .then(() => {
          res.status(204).end();
        });
    }
  };
}

// Gets a list of Recipes
export function index(req, res) {
    Recipe.findAsync({author: req.user._id})
      .then(respondWithResult(res))
      .catch(handleError(res));
}

// Gets a single Recipe from the DB
export function show(req, res) {
  /* @todo: check author */
  Recipe.findByIdAsync(req.params.id)
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Creates a new Recipe in the DB
export function create(req, res) {
  const recipe = req.body;
  recipe.author = req.user._id;
  Recipe.createAsync(recipe)
    .then(respondWithResult(res, 201))
    .catch(handleError(res));
}

// Updates an existing Recipe in the DB
export function update(req, res) {
  if(req.body._id) {
    delete req.body._id;
  }
  Recipe.findByIdAsync(req.params.id)
    .then(handleEntityNotFound(res))
    .then(saveUpdates(req.body))
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Deletes a Recipe from the DB
export function destroy(req, res) {
  Recipe.findByIdAsync(req.params.id)
    .then(handleEntityNotFound(res))
    .then(removeEntity(res))
    .catch(handleError(res));
}
