// client/app/components/profile/DailyIntakeForm.jsx
import React, {PropTypes} from 'react';
import {FormGroup, ControlLabel, FormControl, InputGroup, Button} from 'react-bootstrap';

class DailyIntakeForm extends React.Component {
  constructor(...args) {
    super(...args);
    this.state = {value: this.props.initialValue || 0};
    this.handleChange = this.handleChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.initialValue !== this.state.initialValue) {
      this.setState({value: nextProps.initialValue});
    }
  }

  handleChange({target: {value}}) {
    this.setState({value});
  }

  render() {
    const {onSubmit} = this.props;
    return (
      <form>
        <FormGroup>
          <ControlLabel>Your daily intake (kcal)</ControlLabel>
          <InputGroup>
            <FormControl type="text"
                         value={this.state.value}
                         placeholder="kcal"
                         onChange={this.handleChange} />
            <InputGroup.Button>
              <Button onClick={() => onSubmit(this.state.value)}>Set</Button>
            </InputGroup.Button>
          </InputGroup>
        </FormGroup>
      </form>
    );
  }
}

DailyIntakeForm.propTypes = {
  initialValue: PropTypes.number,
  onSubmit: PropTypes.func,
};

DailyIntakeForm.defaultProps = {
  onSubmit: () => {
  },
};

export default DailyIntakeForm;
