// client/app/components/navbar/Navbar.jsx
import _ from 'lodash'

import React from 'react'
import { IndexLink } from 'react-router'
import { LinkContainer } from 'react-router-bootstrap'
import { Navbar, Nav, NavItem, NavDropdown, MenuItem, Glyphicon } from 'react-bootstrap'

import Omnibox from './Omnibox'
import NotificationDropown from './NotificationDropdown'
import SettingsDropdown from './SettingsDropdown'
import IconLink from './IconLink'

import './navbar.scss'

export default class MainNavbar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: false,
    };
    _.bindAll(this, ['collapseNav', 'toggleExpanded', 'onSearch']);
  }

  collapseNav() {
    this.setState({ expanded: false });
  }

  onSearch(searchTerm) {
    this.collapseNav();
    this.props.onSearch(searchTerm);
  }

  toggleExpanded() {
    this.setState({ expanded: !this.state.expanded });
  }

  render() {
    return (
        <Navbar expanded={this.state.expanded} onToggle={this.toggleExpanded}>
          <Navbar.Header>
            <Navbar.Brand>
              <a href="#">BiteTracker</a>
            </Navbar.Brand>

            <Navbar.Toggle />
          </Navbar.Header>

          <Navbar.Collapse>

            <Navbar.Form pullLeft>
              <Omnibox onSearch={this.onSearch} />
            </Navbar.Form>

            <Nav pullRight>
              {/* Standard navigation links */}
              <LinkContainer to="/recipes">
                <NavItem eventKey={1} onClick={this.collapseNav}>Recipes</NavItem>
              </LinkContainer>
              <LinkContainer to="/log">
                <NavItem eventKey={2} onClick={this.collapseNav}>Food Log</NavItem>
              </LinkContainer>
              <LinkContainer to="/planner">
                <NavItem eventKey={3} onClick={this.collapseNav}>Meal Planner</NavItem>
              </LinkContainer>
              <LinkContainer to="/groceries">
                <NavItem eventKey={4} onClick={this.collapseNav}>Groceries</NavItem>
              </LinkContainer>

              {/* Notification dropdown - only visible on larger screeens */}
              <NotificationDropown eventKey={5} />

              {/* Icon menu items */}
              <IconLink to="/notifications"
                        glyph="bell" text="Notifications"
                        className="navbar-notif-link"
                        eventKey={5} onClick={this.collapseNav} />
              <IconLink to="/profile"
                        glyph="user" text="Profile"
                        eventKey={6} onClick={this.collapseNav} />
              <IconLink to="/settings"
                        glyph="cog" text="Settings"
                        eventKey={7} onClick={this.collapseNav} />

              {/* Settings dropdown - only visible on small devices, replaces menu icons above */}
              <SettingsDropdown eventKey={8} />
            </Nav>

          </Navbar.Collapse>
        </Navbar>
    );
  }
}

MainNavbar.propTypes = {
  onSearch: React.PropTypes.func.isRequired
};
