// client/app/components/navbar/NotificationDropdown.jsx
import React from 'react'
import { NavDropdown, MenuItem, Glyphicon } from 'react-bootstrap'

import NotificationStore from 'client/app/stores/NotificationStore'

export default class NotificationDropdown extends React.Component {
  constructor(props) {
    super(props);
    this.state = NotificationStore.getState();
  }

  componentDidMount() {
    NotificationStore.listen(this.onNotificationsChange)
  }

  componentWillUnmount() {
    NotificationStore.unlisten(this.onNotificationsChange);
  }

  onNotificationsChange(state) {
    this.setState(state);
  }

  getNotificationNodes() {
    let notifications = this.state.notifications;
    return notifications.map((elem, i) =>
        <MenuItem key={elem._id}>{elem.text}</MenuItem>
    );
  }

  render() {
    return (
        <NavDropdown
            {...this.props}
            title={<Glyphicon glyph="bell" aria-label="notifications" />}
            className="navbar-notif-dropdown navbar-icon"
            id="navbar-notif-dropdown"
            aria-hidden={true}>

          {this.getNotificationNodes()}

        </NavDropdown>
    );
  }
}
