// client/app/components/shared/IngredientsAutocomplete.jsx
import React from 'react';
import axios from 'axios';

import Autosuggest from './autosuggest/Autosuggest';
import './IngredientsAutosuggest.scss';

export default class IngredientsAutocomplete extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      ingredient: null,
      ingredients: [],
      message: "msg"
    };

    this.onIngredientSearch = this.onIngredientSearch.bind(this);
    this.onIngredientChange = this.onIngredientChange.bind(this);
  }

  onIngredientSearch(search, page, prev) {
    this.setState({message: "Searching for matching ingredients..."});
    if(search) {
      axios.get('/api/food_search/' + search)
        .then(resp => {
          const fullList = resp.data.list.item;
          this.setState({ingredients: fullList, message: null});
        })
    } else {
      this.setState({ingredients: [], message: null});
    }
  }

  onIngredientChange(selected) {
    if(typeof selected === 'object') {
      this.setState({ingredient: selected});
      this.props.onIngredientSelected(selected);
    }
  }

  render() {
    return (
      <div className="s__iautosuggest">
        <Autosuggest
          datalist={this.state.ingredients}
          datalistMessage={this.state.message}
          placeholder="Choose ingredient..."
          value={this.state.repo}
          //itemAdapter={RepoAdapter.instance}
          itemValuePropName="name"
          searchDebounce={500}
          onSearch={this.onIngredientSearch}
          onSelect={this.onIngredientChange} />
      </div>
    );
  }
}

IngredientsAutocomplete.propTypes = {
  onIngredientSelected: React.PropTypes.func,
};

IngredientsAutocomplete.defaultProps = {
  onIngredientSelected: (ingredient) => {}
};
