// client/app/sources/GroceryListSource.js
import axios from 'axios';
import createAsync from './create-async';
import {arrayOf} from 'normalizr';
import groceryList from './schemas/groceries';
import jsonpatch from 'fast-json-patch';
import GroceryListActions from 'actions/GroceryListActions';

const GROCERY_LISTS_PATH = '/api/grocery_lists';
const GROCERY_LIST_PATH = (id) => `${GROCERY_LISTS_PATH}/${id}`;

const GroceryListSource = {
  fetch: {
    request: () => axios.get(GROCERY_LISTS_PATH),
    success: GroceryListActions.fetchSuccess,
    error: GroceryListActions.fetchFailure,
    normalized: arrayOf(groceryList),
  },
  create: {
    request: dataa => axios.post(GROCERY_LISTS_PATH, data),
    success: GroceryListActions.createSuccess,
    error: GroceryListActions.createFailure,
    normalized: groceryList,
  },
  update: {
    request: (list, newList) => {
      const data = jsonpatch.compare(list, newList);
      return axios.patch(GROCERY_LIST_PATH(list._id), data);
    },
    success: GroceryListActions.updateSuccess,
    error: GroceryListActions.updateFailure,
    normalized: groceryList,
  },
  remove: {
    request: id => axios.delete(GROCERY_LIST_PATH(id)),
    success: GroceryListActions.removeSuccess,
    error: GroceryListActions.removeFailure,
    normalized: groceryList,
  },
};

export default createAsync(GroceryListSource);
