// client/app/sources/GroceryListSource.js
import axios from 'axios';
import GroceryListActions from 'actions/GroceryListActions';

const GROCERY_LISTS_PATH = '/api/grocery_lists';
const GROCERY_LIST_PATH = (id) => `${GROCERY_LISTS_PATH}/${id}`;

const GroceryListSource = {
  fetch: {
    remote() {
      return axios.get(GROCERY_LISTS_PATH);
    },
    local() {
      // TODO: cache API
    },
    success: GroceryListActions.fetchSuccess,
    error: GroceryListActions.fetchFailure,
  },
  create: {
    remote(_, data) {
      return axios.post(GROCERY_LISTS_PATH, data);
    },
    success: GroceryListActions.createSuccess,
    error: GroceryListActions.createFailure,
  },
  update: {
    remote(_, id, data) {
      return axios.patch(GROCERY_LIST_PATH(id), data);
    },
    success: GroceryListActions.updateSuccess,
    error: GroceryListActions.updateFailure,
  },
  remove: {
    remote(_, id){
      return axios.delete(GROCERY_LIST_PATH(id))
    },
    success: GroceryListActions.removeSuccess,
    error: GroceryListActions.removeFailure,
  },
};

export default GroceryListSource;
