// client/app/sources/GroceryListSource.js
import axios from 'axios';
import createAsync from './create-async';
import {arrayOf} from 'normalizr';
import {groceryList} from './schemas/groceries';
import GroceryListActions from 'actions/GroceryListActions';

const GROCERY_LISTS_PATH = '/api/grocery_lists';
const GROCERY_LIST_PATH = (id) => `${GROCERY_LISTS_PATH}/${id}`;

const GroceryListSource = {
  fetch: {
    request: cacheThenNetwork({
      remote: () => axios.get(GROCERY_LISTS_PATH),
      local: () => caches.match(GROCERY_LISTS_PATH),
    }),
    success: GroceryListActions.fetchSuccess,
    error: GroceryListActions.fetchFailure,
    normalized: arrayOf(groceryList),
  },
  create: {
    request: (data) => axios.post(GROCERY_LISTS_PATH, data),
    success: GroceryListActions.createSuccess,
    error: GroceryListActions.createFailure,
    normalized: groceryList,
  },
  update: {
    request: (id, data) => axios.patch(GROCERY_LIST_PATH(id), data),
    success: GroceryListActions.updateSuccess,
    error: GroceryListActions.updateFailure,
    normalized: groceryList,
  },
  remove: {
    request: (id) => axios.delete(GROCERY_LIST_PATH(id)),
    success: GroceryListActions.removeSuccess,
    error: GroceryListActions.removeFailure,
    normalized: groceryList,
  },
};

export default createAsync(GroceryListSource);
