// client/app/components/groceries/aside/GroceriesSidebar.jsx
import assign from 'lodash/assign';
import React from 'react';
import GroceryListActions from 'actions/GroceryListActions';

import AddInput from 'components/shared/form/AddInput';
import GroceryListNav from './GroceryListNav';

import './groceries-sidebar.scss'

export default class GroceryListSidebar extends React.Component {
  constructor(...args) {
    super(...args);

    this.state = {list: {title: ''}};

    this.handleTitleChange = this.handleTitleChange.bind(this);
    this.handleListSubmit = this.handleListSubmit.bind(this);
    this.handleListDelete = this.handleListDelete.bind(this);
  }

  handleTitleChange(title) {
    this.setState({list: assign({}, this.state.list, {title})});
  }

  handleListSubmit() {
    GroceryListActions.addList(this.state.list);
  }

  handleListDelete(listId) {
    GroceryListActions.removeList(listId);
  }

  render() {
    return (
      <aside>
        <AddInput value={this.state.list.title}
                  placeholder="Add list..."
                  onChange={this.handleTitleChange}
                  onSubmit={this.handleListSubmit} />
        <GroceryListNav onDelete={this.handleListDelete} {...this.props} />
      </aside>
    );
  }
}
