// server/api/grocery_list/grocery_list.controller.js
/**
 * Using Rails-like standard naming convention for endpoints.
 * GET     /api/grocery_lists              ->  index
 * POST    /api/grocery_lists              ->  create
 * GET     /api/grocery_lists/:id          ->  show
 * PUT     /api/grocery_lists/:id          ->  upsert
 * PATCH   /api/grocery_lists/:id          ->  patch
 * DELETE  /api/grocery_lists/:id          ->  destroy
 */

'use strict';

import merge from 'lodash/merge';
import GroceryList from './grocery_list.model';

function respondWithResult(res, statusCode) {
  statusCode = statusCode || 200;
  return function (entity) {
    if (entity) {
      return res.status(statusCode).json(entity);
    }
    return null;
  };
}

function saveUpdates(updates) {
  return function(entity) {
    var updated = merge(entity, updates);
    return updated.saveAsync()
      .spread(updated => {
        return updated;
      });
  };
}

function removeEntity(res) {
  return function (entity) {
    if (entity) {
      return entity.remove()
        .then(() => {
          res.status(204).end();
        });
    }
  };
}

function handleEntityNotFound(res) {
  return function (entity) {
    if (!entity) {
      res.status(404).end();
      return null;
    }
    return entity;
  };
}

function handleError(res, statusCode) {
  statusCode = statusCode || 500;
  return function (err) {
    res.status(statusCode).send(err);
  };
}

// Gets a list of GroceryLists
export function index(req, res) {
  return GroceryList.find().exec()
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Gets a single GroceryList from the DB
export function show(req, res) {
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Creates a new GroceryList in the DB
export function create(req, res) {
  return GroceryList.create(req.body)
    .then(respondWithResult(res, 201))
    .catch(handleError(res));
}

// Updates an existing GroceryList in the DB
export function update(req, res) {
  if (req.body._id) {
    delete req.body._id;
  }
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(saveUpdates(req.body))
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Deletes a GroceryList from the DB
export function destroy(req, res) {
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(removeEntity(res))
    .catch(handleError(res));
}
