// server/api/grocery_list/grocery_list.controller.js
/**
 * Using Rails-like standard naming convention for endpoints.
 * GET     /api/grocery_lists              ->  index
 * POST    /api/grocery_lists              ->  create
 * GET     /api/grocery_lists/:id          ->  show
 * PUT     /api/grocery_lists/:id          ->  upsert
 * PATCH   /api/grocery_lists/:id          ->  patch
 * DELETE  /api/grocery_lists/:id          ->  destroy
 */

'use strict';

import jsonpatch from 'fast-json-patch';
import GroceryList from './grocery_list.model';

function respondWithResult(res, statusCode) {
  statusCode = statusCode || 200;
  return function (entity) {
    if (entity) {
      return res.status(statusCode).json(entity);
    }
    return null;
  };
}

function patchUpdates(patches) {
  return function (entity) {
    try {
      jsonpatch.apply(entity, patches, /*validate*/ true);
    } catch (err) {
      return Promise.reject(err);
    }

    return entity.save();
  };
}

function removeEntity() {
  return function (entity) {
    if (entity) {
      return entity.remove();
    }
  };
}

function handleEntityNotFound(res) {
  return function (entity) {
    if (!entity) {
      res.status(404).end();
      return null;
    }
    return entity;
  };
}

function handleError(res, statusCode) {
  statusCode = statusCode || 500;
  return function (err) {
    res.status(statusCode).send(err);
  };
}

// Gets a list of GroceryLists
export function index(req, res) {
  return GroceryList.find().exec()
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Gets a single GroceryList from the DB
export function show(req, res) {
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Creates a new GroceryList in the DB
export function create(req, res) {
  return GroceryList.create(req.body)
    .then(respondWithResult(res, 201))
    .catch(handleError(res));
}

// Upserts the given GroceryList in the DB at the specified ID
export function upsert(req, res) {
  if (req.body._id) {
    delete req.body._id;
  }
  return GroceryList
    .findOneAndUpdate({_id: req.params.id}, req.body, {upsert: true, setDefaultsOnInsert: true, runValidators: true})
    .exec()
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Updates an existing GroceryList in the DB
export function patch(req, res) {
  if (req.body._id) {
    delete req.body._id;
  }
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(patchUpdates(req.body))
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Deletes a GroceryList from the DB
export function destroy(req, res) {
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(removeEntity())
    .then(respondWithResult(res))
    .catch(handleError(res));
}
