// server/api/grocery_list/grocery_list.controller.js
/**
 * GET     /api/grocery_lists              ->  index
 * POST    /api/grocery_lists              ->  create
 * GET     /api/grocery_lists/:id          ->  show
 * PUT     /api/grocery_lists/:id          ->  upsert
 * PATCH   /api/grocery_lists/:id          ->  patch
 * DELETE  /api/grocery_lists/:id          ->  destroy
 */

import {respondWithResult, patchUpdates, removeEntity, handleEntityNotFound, handleError} from 'server/api/common';
import GroceryList from './grocery_list.model';

// Gets a list of GroceryLists
export function index(req, res) {
  return GroceryList.find().exec()
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Gets a single GroceryList from the DB
export function show(req, res) {
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Creates a new GroceryList in the DB
export function create(req, res) {
  return GroceryList.create(req.body)
    .then(respondWithResult(res, 201))
    .catch(handleError(res));
}

// Upserts the given GroceryList in the DB at the specified ID
export function upsert(req, res) {
  if (req.body._id) {
    delete req.body._id;
  }
  return GroceryList
    .findOneAndUpdate({_id: req.params.id}, req.body, {upsert: true, setDefaultsOnInsert: true, runValidators: true})
    .exec()
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Updates an existing GroceryList in the DB
export function patch(req, res) {
  if (req.body._id) {
    delete req.body._id;
  }
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(patchUpdates(req.body))
    .then(respondWithResult(res))
    .catch(handleError(res));
}

// Deletes a GroceryList from the DB
export function destroy(req, res) {
  return GroceryList.findById(req.params.id).exec()
    .then(handleEntityNotFound(res))
    .then(removeEntity())
    .then(respondWithResult(res))
    .catch(handleError(res));
}
